# 결측치를 좀더 꼼꼼하게 대치

## 1. Data Load

In [336]:

import pandas as pd
import numpy as np

train=pd.read_csv(r"C:\Users\USER\Desktop\캐글 모델링\boston house\data\train.csv", index_col='Id')
test=pd.read_csv(r"C:\Users\USER\Desktop\캐글 모델링\boston house\data\test.csv", index_col='Id')
# 프로파일 및 eda 목적용 데이터 프레임
data=pd.concat([train, test], sort=False)


## 2. 결측치 확인

In [337]:
missing_col=list(data.columns[data.isnull().any()])
missing_sorted=data[missing_col].isnull().sum().sort_values(ascending=False)

#결측치가 '없음'을 의미하는 특성들의 결측치는 'nan'으로 채움
non_col=[
    "Alley", "BsmtQual", "BsmtCond", "BsmtExposure", "BsmtFinType1",
    "BsmtFinType2", "FireplaceQu", "GarageType", "GarageFinish",
    "GarageQual", "GarageCond", "PoolQC", "Fence", "MiscFeature"
]
for i in non_col:
    data[i]=data[i].fillna('nan')

for i in non_col:
     missing_col.remove(i)

In [338]:
missing_col=list(data.columns[data.isnull().any()])
missing_sorted=data[missing_col].isnull().sum().sort_values(ascending=False)
print(missing_col)
print(missing_sorted)

['MSZoning', 'LotFrontage', 'Utilities', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Electrical', 'BsmtFullBath', 'BsmtHalfBath', 'KitchenQual', 'Functional', 'GarageYrBlt', 'GarageCars', 'GarageArea', 'SaleType', 'SalePrice']
SalePrice       1459
LotFrontage      486
GarageYrBlt      159
MasVnrType        24
MasVnrArea        23
MSZoning           4
BsmtFullBath       2
BsmtHalfBath       2
Functional         2
Utilities          2
Exterior2nd        1
BsmtFinSF1         1
BsmtFinSF2         1
BsmtUnfSF          1
Exterior1st        1
Electrical         1
KitchenQual        1
GarageCars         1
GarageArea         1
SaleType           1
TotalBsmtSF        1
dtype: int64


In [339]:
# import pandas_profiling
# pr=data[missing_col].profile_report()
# pr.to_file(f'./결측치 컬럼.html')

LotFrontage      486

GarageYrBlt      159

MasVnrType        24

MasVnrArea        23

이 4가지를 제외하면 결측치가 1,2,4개로

위의 4가지의 결측치 처리가 가장 중요함

최빈값
MSZoning
Utilities
Exterior1st
Exterior2nd
BsmtFinSF1
BsmtFinSF2
BsmtUnfSF
Electrical
BsmtFullBath
BsmtHalfBath
KitchenQual
Functional
GarageCars
SaleType

보류
-MasVnrArea
-LotFrontage
-GarageYrBlt


평균
TotalBsmtSF
GarageArea

비율
-MasVnrType


1. 최빈값 대체

In [340]:
mode_list=['MSZoning', 'Utilities', 'Exterior1st', 'Exterior2nd', 
                'BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF', 'Electrical', 
                'BsmtFullBath', 'BsmtHalfBath', 'KitchenQual', 'Functional', 
                'GarageCars', 'SaleType']
for i in mode_list:
    data[i] = data[i].fillna(data[i].mode().iloc[0])

2. 평균값 대체

In [341]:
mean_list=['TotalBsmtSF','GarageArea']
for i in mean_list:
    data[i] = data[i].fillna(data[i].mean())

In [342]:
data["MasVnrArea"] = data["MasVnrArea"].fillna(0)

In [343]:
#없음 = "None"
data["MasVnrType"] = data["MasVnrType"].fillna("None")

In [344]:
# 이웃의 중앙값
data["LotFrontage"] = data.groupby("Neighborhood")["LotFrontage"].transform(lambda x: x.fillna(x.median()))

In [345]:
#차고 건설연도에 결측값 -> 없다고 판단 (어차피 원핫인코딩할것이기때문에 nan으로 채움)
data["GarageYrBlt"] = data["GarageYrBlt"].fillna('Nan')

In [346]:
missing_col=list(data.columns[data.isnull().any()])
missing_sorted=data[missing_col].isnull().sum().sort_values(ascending=False)
print(missing_col)

['SalePrice']
